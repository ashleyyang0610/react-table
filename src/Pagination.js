import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './styles/index.styl';

const Pagination = ({
    className,
    onPageChange,
    page,
    pageLength,
    prevPageRenderer,
    nextPageRenderer,
    lastPageRenderer,
    firstPageRenderer,
    ...props
}) => {
    const btnConfig = [
        {
            key: 'first-page',
            isDisabled: page === 1,
            onClick: () => {
                onPageChange(1);
            },
            render: firstPageRenderer()
        },
        {
            key: 'prev-page',
            isDisabled: page === 1,
            onClick: () => {
                onPageChange(page - 1);
            },
            render: prevPageRenderer()
        },
        {
            key: 'next-page',
            isDisabled: page === pageLength,
            onClick: () => {
                onPageChange(page + 1);
            },
            render: nextPageRenderer()
        },
        {
            key: 'last-page',
            isDisabled: page === pageLength,
            onClick: () => {
                onPageChange(pageLength);
            },
            render: lastPageRenderer()
        }
    ];

    return (
        <div {...props} className={cx(styles.pagination, className)}>
            {btnConfig.map(config => (
                <button
                    key={config.key}
                    type="button"
                    disabled={config.isDisabled}
                    onClick={config.onClick}
                >
                    {config.render}
                </button>
            ))}
        </div>
    );
};

Pagination.defaultProps = {
    prevPageRenderer: () => 'Prev',
    nextPageRenderer: () => 'Next',
    lastPageRenderer: () => 'Last',
    firstPageRenderer: () => 'First'
};

Pagination.propTypes = {
    /**
     * Action dispatched when page onchnage.
     */
    onPageChange: PropTypes.func.isRequired,
    /**
     * Current page.
     */
    page: PropTypes.number.isRequired,
    /**
     * Total page length.
     */
    pageLength: PropTypes.number.isRequired,
    /**
     * Content renderer in "previous" page button.
     */
    prevPageRenderer: PropTypes.func,
    /**
     * Content renderer in "next page" button.
     */
    nextPageRenderer: PropTypes.func,
    /**
     * Content renderer in "last page" button.
     */
    lastPageRenderer: PropTypes.func,
    /**
     * Content renderer in "first page" button.
     */
    firstPageRenderer: PropTypes.func
};

export default Pagination;
