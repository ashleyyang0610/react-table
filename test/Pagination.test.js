import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Pagination from '../src/Pagination';

describe('Pagination', () => {
    let pagination;
    let onPageChange;

    describe('with default settings', () => {
        beforeEach(() => {
            onPageChange = jest.fn();
            pagination = shallow(
                <Pagination
                    page={1}
                    onPageChange={onPageChange}
                    pageLength={8}
                />
            );
        });

        test('should render correctly', () => {
            expect(toJSON(pagination)).toMatchSnapshot();
        });

        test('onPageChange props should be triggered and the page value should be updated when "first" button is click', () => {
            pagination
                .find('button')
                .at(0)
                .simulate('click');
            pagination.update();
            expect(toJSON(pagination)).toMatchSnapshot();
            expect(onPageChange.mock.calls.length).toBe(1);
        });

        test('onPageChange props should be triggered and the page value should be updated when "prev" button is click', () => {
            pagination
                .find('button')
                .at(1)
                .simulate('click');
            pagination.update();
            expect(toJSON(pagination)).toMatchSnapshot();
            expect(onPageChange.mock.calls.length).toBe(1);
        });

        test('onPageChange props should be triggered and the page value should be updated when "next" button is click', () => {
            pagination
                .find('button')
                .at(2)
                .simulate('click');
            pagination.update();
            expect(toJSON(pagination)).toMatchSnapshot();
            expect(onPageChange.mock.calls.length).toBe(1);
        });

        test('onPageChange props should be triggered and the page value should be updated when "last" button is click', () => {
            pagination
                .find('button')
                .at(3)
                .simulate('click');
            pagination.update();
            expect(toJSON(pagination)).toMatchSnapshot();
            expect(onPageChange.mock.calls.length).toBe(1);
        });
    });
});
