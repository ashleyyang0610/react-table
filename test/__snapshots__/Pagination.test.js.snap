// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination with default settings onPageChange props should be triggered and the page value should be updated when "first" button is click 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    key="first-page"
    onClick={[Function]}
    type="button"
  >
    First
  </button>
  <button
    disabled={true}
    key="prev-page"
    onClick={[Function]}
    type="button"
  >
    Prev
  </button>
  <button
    disabled={false}
    key="next-page"
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    disabled={false}
    key="last-page"
    onClick={[Function]}
    type="button"
  >
    Last
  </button>
</div>
`;

exports[`Pagination with default settings onPageChange props should be triggered and the page value should be updated when "last" button is click 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    key="first-page"
    onClick={[Function]}
    type="button"
  >
    First
  </button>
  <button
    disabled={true}
    key="prev-page"
    onClick={[Function]}
    type="button"
  >
    Prev
  </button>
  <button
    disabled={false}
    key="next-page"
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    disabled={false}
    key="last-page"
    onClick={[Function]}
    type="button"
  >
    Last
  </button>
</div>
`;

exports[`Pagination with default settings onPageChange props should be triggered and the page value should be updated when "next" button is click 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    key="first-page"
    onClick={[Function]}
    type="button"
  >
    First
  </button>
  <button
    disabled={true}
    key="prev-page"
    onClick={[Function]}
    type="button"
  >
    Prev
  </button>
  <button
    disabled={false}
    key="next-page"
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    disabled={false}
    key="last-page"
    onClick={[Function]}
    type="button"
  >
    Last
  </button>
</div>
`;

exports[`Pagination with default settings onPageChange props should be triggered and the page value should be updated when "prev" button is click 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    key="first-page"
    onClick={[Function]}
    type="button"
  >
    First
  </button>
  <button
    disabled={true}
    key="prev-page"
    onClick={[Function]}
    type="button"
  >
    Prev
  </button>
  <button
    disabled={false}
    key="next-page"
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    disabled={false}
    key="last-page"
    onClick={[Function]}
    type="button"
  >
    Last
  </button>
</div>
`;

exports[`Pagination with default settings should render correctly 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    key="first-page"
    onClick={[Function]}
    type="button"
  >
    First
  </button>
  <button
    disabled={true}
    key="prev-page"
    onClick={[Function]}
    type="button"
  >
    Prev
  </button>
  <button
    disabled={false}
    key="next-page"
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
  <button
    disabled={false}
    key="last-page"
    onClick={[Function]}
    type="button"
  >
    Last
  </button>
</div>
`;
